### PR TITLE
PDP 1185: Fourth batch of reworks

### PR DESCRIPTION
**Pull request by danibishop**
This PR improves readability and fixes some minor typos and bugs across some features

### CODE REVIEW

File: lib/sum_of_depository_balances_latest/main.py [hunk 1]
```diff
@@ -27,40 +27,50 @@ async def get_sum_of_depository_balances_latest(
...
-    for ix, inst_list in enumerate(r):
+    balances_per_institution = await asyncio.gather(*inst_coroutines)
+
+    # STEP 3: We flatten the lists of balances into a single list of balances
+    balances_flattened = []
+    for ix, inst_list in enumerate(balances_per_institution):
         institution_id = institutions[ix].institution_id
-        records_list.extend([dict(transaction, institution_id=institution_id) for transaction in inst_list])
+        for balance in inst_list:
+            balances_flattened.append(dict(balance, institution_id=institution_id))
```
Comment (JaniceCanedo) [inline lib/sum_of_depository_balances_latest/main.py:60]: same comment as before on dict. Does this translate well to other languages?

File: lib/sum_of_depository_balances_latest/main.py [hunk 2]
```diff
@@ -27,40 +27,50 @@ async def get_sum_of_depository_balances_latest(
...
     
-    # Here we sort by timestamp so latest balances are on top
-    records_list = sorted(records_list, key=lambda x: x["ts"], reverse=True)
+    # STEP 5: Here we sort by timestamp so latest balances are on top
+    balances_flattened = sorted(balances_flattened, key=lambda x: x["ts"], reverse=True)
```
Comment (JaniceCanedo) [inline lib/sum_of_depository_balances_latest/main.py:63]: same comment as before. I can't think of a better way that's more readable. Can you or do you feel this is pretty straight forward?

File: lib/sum_of_loan_repayments/main.py [hunk 3]
```diff
@@ -30,33 +30,38 @@ async def get_sum_of_loan_repayments(
...
-    r = await asyncio.gather(*inst_coroutines)
-    for inst_lst in r:
-        record_list.extend([dict(transaction) for transaction in inst_lst])
+
+    transactions_per_institution = await asyncio.gather(*inst_coroutines)
+    
+    # STEP 3: We flatten the lists of transactions into a single list of transactions
+    transactions_flattened = []
+    for inst_lst in transactions_per_institution:
+        transactions_flattened.extend([dict(transaction) for transaction in inst_lst])
```
Comment (JaniceCanedo) [inline lib/sum_of_loan_repayments/main.py:60]: There a list comprehension here.

**Review by JaniceCanedo — CHANGES_REQUESTED** [review]
1 list comprehension to for loop. Otherwise a couple questions as repeated from the other PR.

File: lib/sum_of_depository_balances_latest/main.py (hunk 1 above)
Comment (danibishop) [reply lib/sum_of_depository_balances_latest/main.py:60]: I think we can leave these here and maybe refactor all these uses of the `dict`cast (which are not so many at the end). If I find a nice way to express this I will post a new PR after current ones are merged.

File: lib/sum_of_depository_balances_latest/main.py (hunk 2 above)
Comment (danibishop) [reply lib/sum_of_depository_balances_latest/main.py:63]: As I commented in previous PRs, I think most modern languages already have a lambda-like sort base function so this is not so just-python stuff. I will add this to potential final refactors.

File: lib/sum_of_loan_balances_latest/main.py [hunk 4]
```diff
@@ -26,44 +26,54 @@ async def get_sum_of_loan_balances_latest(
     Returns:
         The latest balance summed across all loan accounts
     """
-
+    
+    # STEP 1: get a list of all institutions for the user
     institutions = await api_client.institutions.get(user_uuid=user_uuid)
     utc_starttime = utc_time - cutoff_interval
```
[the remark this answers is not in the document]
Comment (danibishop) [reply lib/sum_of_loan_balances_latest/main.py:32]: This has been addressed in next PRs.

File: lib/sum_of_loan_repayments/main.py (hunk 3 above)
Comment (danibishop) [reply lib/sum_of_loan_repayments/main.py:60]: Addressed in next PRs

File: lib/sum_of_debits/main.py [hunk 5]
```diff
@@ -29,42 +29,39 @@ async def get_sum_of_debits(
...
+    )
 
-    # now we sum the transaction amounts in each time window
-    amount = None
+    transactions_per_institution = await asyncio.gather(*inst_coroutines)
 
-    for r in record_list:
-        if r["impact"] == "DEBIT":
-            if r["ts"] >= utc_starttime.timestamp() and r["ts"] < utc_endtime.timestamp():
-                amount = _update_total(r["amount"], amount)
```
Comment (frankie-ug) [inline lib/sum_of_debits/main.py:67]: I think we forget to delete `_update_total` function

**Review by frankie-ug — APPROVED** [review]
LGTM 👍 Just 1 minor comment

File: lib/sum_of_debits/main.py [hunk 6]
```diff
@@ -29,42 +29,39 @@ async def get_sum_of_debits(
...
+    # STEP 3: now we sum up all the amounts of debit transactions for each institution
+    amount = 0
+    for transactions in transactions_per_institution:
+        for transaction in transactions:
+            if transaction.impact == "DEBIT" and transaction.amount is not None:
```
[the remark this answers is not in the document]
Comment (danibishop) [reply lib/sum_of_debits/main.py:63]: Sure, but that implies that the user reading the code knows how the automatic casting of `transaction.amount` works. I think we prefer being explicit here, but thanks for the idea!

**Pull request merged**